index.js

```diff
@@ -199,64 +201,74 @@ Dat.prototype.download = function (cb) {
     if (err) return cb(err)
     self.live = archive.live
     self.db.put('!dat!key', archive.key.toString('hex'))
+    updateTotalStats() //  Call once for downloads previously started
 
     archive.metadata.once('download-finished', updateTotalStats)
 
+    archive.metadata.on('update', function () {
+      updateTotalStats() // Updates total on live sync
+      self.emit('archive-updated')
+    })
+
     archive.content.on('download-finished', function () {
-      if (self.stats.bytesTotal === 0) updateTotalStats() // TODO: why is this getting here with 0
+      updateTotalStats()
       self.emit('download-finished')
     })
 
-    each(archive.list({live: archive.live}), function (data, next) {
-      var startBytes = self.stats.bytesProgress
-      archive.download(data, function (err) {
+    archive.on('download', function (data) {
+      self.stats.blocksProgress++
+      self.stats.bytesDown += data.length
+      self.emit('download', data)
+    })
+
+    archive.on('upload', function (data) {
+      self.stats.bytesUp += data.length
+      self.emit('upload', data)
+    })
+
+    each(archive.list({live: archive.live}), function (entry, next) {
+      if (archive.isEntryDownloaded(entry)) {
+        self.stats.blocksProgress += entry.blocks
+        return entryDone()
+      }
+      var downloaded = archive.countDownloadedBlocks(entry)
+      self.stats.blocksProgress += downloaded
+      archive.download(entry, function (err) {
+        // Other blocks are added to stats.blocksProgress with archive.on('download')
         if (err) return cb(err)
-        if (startBytes === self.stats.bytesProgress) {
-          // TODO: better way to measure progress with existing files
-          self.stats.bytesProgress += data.length
-        }
-        if (data.type === 'file') {
+        entryDone()
+      })
+
+      function entryDone () {
```

Comment:
The entry done function will be created for every entry. It will be more efficient to put this outside the loop and have it take the entry parameter.